Cgen declares variable in expressions

### Describe the bug

<title>

### Reproduction Steps

```v
pub struct Client {
pub:
    token      string
    intents    int
mut:
    ready    bool
    sequence ?int
}

enum Intents { foo }

pub type IntentsOrInt = Intents | int

@[params]
pub struct BotConfig {
pub:
    intents IntentsOrInt
}

pub fn bot(token string, config BotConfig) Client {
    return Client{
        token: 'Bot ${token}'
        intents: match config.intents {
            Intents { int(config.intents) }
            int { config.intents }
        }
    }
}

fn main() {}
```

### Expected Behavior

Successful compile

### Current Behavior

```
==================
C:/Users/mclr/AppData/Local/Temp/v_0/repro.10728893038999124359.tmp.c:6863: warning: implicit declaration of function 'tcc_backtrace'
C:/Users/mclr/AppData/Local/Temp/v_0/repro.10728893038999124359.tmp.c:12832: error: ')' expected (got "_option_int")
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.3 08189d6

### Environment details (OS name and version, etc.)

```v
V full version: V 0.4.3 140c838.08189d6
OS: windows, Microsoft Windows 11 Pro v22000 64-bit
Processor: 12 cpus, 64bit, little endian, 

getwd: D:\Games\Proekti\V\interactions
vexe: D:\Games\Proekti\V\v\v.exe
vexe mtime: 2023-12-09 09:55:27

vroot: OK, value: D:\Games\Proekti\V\v
VMODULES: OK, value: C:\Users\mclr\.vmodules
VTMP: OK, value: C:\Users\mclr\AppData\Local\Temp\v_0

Git version: git version 2.37.0.windows.1
Git vroot status: weekly.2023.48-76-g08189d64
.git/config present: true

CC version:
thirdparty/tcc status: thirdparty-windows-amd64 e90c2620
```
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/struct.v
+++ b/vlib/v/gen/c/struct.v
@@ -378,8 +378,7 @@ fn (mut g Gen) zero_struct_field(field ast.StructField) bool {
 		}
 
 		if field.default_expr is ast.None {
-			tmp_var := g.new_tmp_var()
-			g.expr_with_tmp_var(ast.None{}, ast.none_type, field.typ, tmp_var)
+			g.gen_option_error(field.typ, ast.None{})
 			return true
 		} else if field.typ.has_flag(.option) {
 			tmp_var := g.new_tmp_var()
@@ -394,8 +393,7 @@ fn (mut g Gen) zero_struct_field(field ast.StructField) bool {
 		}
 		g.expr(field.default_expr)
 	} else if field.typ.has_flag(.option) {
-		tmp_var := g.new_tmp_var()
-		g.expr_with_tmp_var(ast.None{}, ast.none_type, field.typ, tmp_var)
+		g.gen_option_error(field.typ, ast.None{})
 		return true
 	} else if sym.info is ast.ArrayFixed {
 		g.write('{')

--- a/vlib/v/tests/option_struct_field_init_test.v
+++ b/vlib/v/tests/option_struct_field_init_test.v
@@ -1,0 +1,34 @@
+pub struct Client {
+pub:
+	token   string
+	intents int
+mut:
+	ready    bool
+	sequence ?int
+}
+
+enum Intents {
+	foo
+}
+
+pub type IntentsOrInt = Intents | int
+
+@[params]
+pub struct BotConfig {
+pub:
+	intents IntentsOrInt
+}
+
+pub fn bot(token string, config BotConfig) Client {
+	return Client{
+		token: 'Bot ${token}'
+		intents: match config.intents {
+			Intents { int(config.intents) }
+			int { config.intents }
+		}
+	}
+}
+
+fn test_main() {
+	assert true
+}

